### DISCUSSION TITLE
SIGSEGV with worker threads

### DISCUSSION THREAD

**Discussion by Zajozor** [avajs/ava#3162] [Q&A] [answered]
* What you're trying to do

I have a typescript project, I'm using tsc to precompile the files, then running ava on them. The generated js is commonJS.
I wrote only a few tests so far and noticed that running with worker threads causes a SIGSEGV.

---
* What happened
```
❯ yarn ava --config tests_ava/ava.config.js
...
error Command failed with signal "SIGSEGV".
```

but when I do 
```
❯ yarn ava --config tests_ava/ava.config.js --no-worker-threads
...
  3 tests passed
✨  Done in 1.91s.
```

---
* What you expected to happen

I'd expect not to get a SIGSEGV when using worker threads.

---
ava.config.js:
```
module.exports = {
  files: ['tests_ava/**', '!tests_ava/ava.config.js'],
  typescript: {
    compile: 'tsc',
    rewritePaths: {
      '@common/': 'dist_test/common/src/',
    },
  },
  failFast: true,
  failWithoutAssertions: false,
}
```

---

I understand that the above is not a minimal reproducible example (MRE) as well as the fact that the issue is likely with some of the modules my project uses.
But would you be so kind as to provide me with some guidance on what to look for, or how to produce an MRE?
I'll happily spend more time digging into it, but I feel a bit lost right now.

1. I specifically noticed that this happens when testing code which interacts with some native modules. Why would this problem not happen with separate processes as well?
2. When the error happens, it looks like all tests pass, all output is written, and the SIGSEGV happens completely at the end, as if it was during some teardown procedure. Is that something I can use when trying to nail this down?
3. The codebase is pretty big - several native and non-native modules - can you suggest something, eg. some node flags I could use to trace this down to a more specific culprit?

Thanks for your time :) !

**Comment by novemberborn** [comment, marked as answer]
Most likely some of the native module code you're interacting with doesn't play well with worker threads. That's an issue with those modules and/or Node.js, not AVA.

I'd try and interact with those modules outside of AVA, using the https://nodejs.org/api/worker_threads.html API directly. 

In the mean time, you're best off running without worker threads, as you already discovered.

**Reply by Zajozor** [reply]
Thanks for the reply :) !

This gives me some clues on what to look into.

One more question if I may - is there any obvious downside to not using the worker threads? ie. is the new process spawning causing some overhead, or are some features just not supported?
I guess since AVA decided to default to worker threads there is some ingerent benefit to them 🤔 

(I tried to do a bit of issue search on this and could not find a distinct answer)

**Reply by novemberborn** [reply]
Worker threads should be a little faster. And the shared workers functionality is only available with worker threads, if that's something you rely on.

**Reply by Zajozor** [reply]
undestood, thanks :) 
marking as an answer and consider this closed.

If I ever need worker threads, and I'll drill down to whatever native package is causing this, I'll post it back here for anyone who finds this on the internet.

For reference, it's probably one of these in our case:

- https://github.com/prisma/prisma (quite likely it's this, based on past experience)
- https://github.com/bripkens/event-loop-stats
- https://www.npmjs.com/package/gc-stats
- https://www.npmjs.com/package/bcrypt
